app/javascript/src/ActiveDocs/ThreeScaleApiDocs.ts

```diff
@@ -17,6 +20,96 @@ const appendSwaggerDiv = (container: HTMLElement, id: string): void => {
   container.appendChild(div)
 }
 
+/**
+ * when using Record notation, the following error is thrown:
+ * 'TS2456: Type alias 'BodyValue' circularly references itself.'
+ */
+// eslint-disable-next-line @typescript-eslint/consistent-indexed-object-style
+type BodyValue = boolean | number | string | { [key: string]: BodyValue }
+
+/**
+ * Transforms an object into form data representation, also URL-encoding the values,
+ * Example:
+ * {
+ *   a_string: 'hello',
+ *   an_array: [
+ *     { first: 1 },
+ *     { second: 1, extra_param: 'with whitespace'}
+ *   ]
+ * }
+ * =>
+ * {
+ *   a_string: 'hello',
+ *   'an_array[0][first]': '1',
+ *   'an_array[1][second]': '1',
+ *   'an_array[1][extra_param]': 'with%20whitespace'
+ * }
+ * @param object
+ */
+export const objectToFormData = (object: BodyValue): Record<string, boolean | number | string> => {
+  const buildFormData = (formData: Record<string, boolean | number | string>, data: BodyValue, parentKey?: string) => {
+    if (data && typeof data === 'object') {
+      Object.keys(data).forEach((key: string) => {
+        buildFormData(formData, data[key], parentKey ? `${parentKey}[${key}]` : key)
+      })
+    } else {
+      if (parentKey) {
+        formData[parentKey] = data ? data : ''
+      }
+    }
+  }
+  const formData = {}
+  buildFormData(formData, object)
+  return formData
+}
+
+/**
+ * Transforms the request body of the Service Management API Report, as
+ * SwaggerUI (or rather swagger-js) does not serialize arrays of objects properly
+ * The hack is to process the request body in two steps:
+ * 1. normalize the 'transactions' array, as its elements may be either an object (if the value is taken
+ *    from the example in the spec), or as a serialized JSON (if the field is changed manually)
+ * 2. "flatten" the objects by transforming them into form-data structure with the entries like
+ *    'transactions[0][app_id]': 'example'
+ *    'transactions[0][usage][hits]': 1
+ * @param body BackendApiReportBody
+ */
+export const transformReportRequestBody = (body: BackendApiReportBody): Record<string, boolean | number | string> => {
+  if (Array.isArray(body.transactions)) {
+    body.transactions = body.transactions.map(transaction => {
+      switch (typeof transaction) {
+        case 'object':
+          return transaction
+        case 'string':
+          try {
+            return JSON.parse(transaction) as BackendApiTransaction
+          } catch (error: unknown) {
+            return null
+          }
+        default:
+          return null
+      }
+    }).filter(element => element != null) as BackendApiTransaction[]
+  }
+  return objectToFormData(body as BodyValue)
```

Comment:
`body` is parsed to `BodyValue` even though it's the only place where `objectToFormData` is used. Why not using `BackendApiReportBody`?

Reply:
My idea was to make `objectToFormData` generic enough so it could potentially be reusable in case we need to do the same serialization for other endpoints - I believe it's not needed at this point, but IMO it doesn't make sense to make it more specific either.

And also in any case when we enter the recursive function we'd need to cast to a more generic class...

Reply:
I prefer naming interfaces with capital I, e.g. `IBodyValue` 🤓 BUT it's perfect like that 🙌🏻